### PR TITLE
[SYNCOPE-1716] fixed membership hashmap multi-threaded problem with concurrent pull

### PR DESCRIPTION
**Pull request by flyueliu**

### CODE REVIEW

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java [hunk 1]
```diff
@@ -141,7 +141,7 @@ protected List<Object> getMembAttrValues(final SyncDelta delta, final Connector
      */
     @Transactional(readOnly = true)
     @Override
-    public void beforeUpdate(
+    public synchronized void beforeUpdate(
```
Comment (ilgrosso) [inline core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java:144]: please find a different way to solve the issue which does not involve synchronizing the whole methods

**Review by ilgrosso — CHANGES_REQUESTED** [review]

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java (hunk 1 above)
Comment (flyueliu) [reply core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java]: yes, use ConcurrentHashMap is better way.

**Comment by flyueliu** [comment]
> Thanks for your PR @flyueliu Please ensure to fill, sign and send your [ICLA](https://www.apache.org/licenses/icla.pdf) before we can accept your contributions

ok, i have send a email to secretary@apache.org

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java [hunk 2]
```diff
@@ -155,8 +156,10 @@ public void beforeUpdate(
         groupDAO.findUMemberships(groupDAO.find(entity.getKey())).forEach(uMembership -> {
             Set<String> memb = membershipsBefore.get(uMembership.getLeftEnd().getKey());
             if (memb == null) {
-                memb = new HashSet<>();
-                membershipsBefore.put(uMembership.getLeftEnd().getKey(), memb);
+                memb = Collections.synchronizedSet(new HashSet<>());
```
Comment (ilgrosso) [inline core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java:159]: This snippet should be replaced by using `ConcurrentHashMap#computeIfAbsent`, which involves declaring both maps as `ConcurrentMap` - see https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#computeIfAbsent-K-java.util.function.Function-

e.g.

```java
            Set<String> memb = membershipsBefore.computeIfAbsent(
                    uMembership.getLeftEnd().getKey(),
                    k -> Collections.synchronizedSet(new HashSet<>()));
            memb.add(entity.getKey());
```

**Review by ilgrosso — CHANGES_REQUESTED** [review]

**Comment by ilgrosso** [comment]
> > Thanks for your PR @flyueliu Please ensure to fill, sign and send your [ICLA](https://www.apache.org/licenses/icla.pdf) before we can accept your contributions
> 
> ok, i have send a email to [secretary@apache.org](mailto:secretary@apache.org)

Did you happen to ask the Syncope project to be notified? I can't seem to find anything yet in the records.

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java (hunk 2 above)
Comment (flyueliu) [reply core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java]: Thank your review, let me learned new API

**Comment by flyueliu** [comment]
> > > Thanks for your PR @flyueliu Please ensure to fill, sign and send your [ICLA](https://www.apache.org/licenses/icla.pdf) before we can accept your contributions
> > 
> > 
> > ok, i have send a email to [secretary@apache.org](mailto:secretary@apache.org)
> 
> Did you happen to ask the Syncope project to be notified? I can't seem to find anything yet in the records.

II did not fill in the notify project in my previous email, so I have now sent a new one

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java [hunk 3]
```diff
@@ -153,11 +154,8 @@ public void beforeUpdate(
         }
 
         groupDAO.findUMemberships(groupDAO.find(entity.getKey())).forEach(uMembership -> {
-            Set<String> memb = membershipsBefore.get(uMembership.getLeftEnd().getKey());
-            if (memb == null) {
-                memb = new HashSet<>();
-                membershipsBefore.put(uMembership.getLeftEnd().getKey(), memb);
-            }
+            Set<String> memb = membershipsBefore.computeIfAbsent(uMembership.getLeftEnd().getKey(),
```
Comment (ilgrosso) [inline core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/pushpull/LDAPMembershipPullActions.java:157]: Not fundamental, but can you keep the same formatting as the other change below?

**Review by ilgrosso — CHANGES_REQUESTED** [review]

**Comment by ilgrosso** [comment]
@flyueliu there is [one unit test failure](https://github.com/apache/syncope/actions/runs/7150630020/job/19475613454?pr=571#step:8:5890), please fix:

```
Error:    LDAPMembershipPullActionsTest.after:209 » NullPointer Cannot invoke "java.util.Set.add(Object)" because "memb" is null
```

**Review by ilgrosso — APPROVED** [review]

**Comment by ilgrosso** [comment]
Thank you very much @flyueliu 
I've cherry-picked this PR to branch `3_0_X` as 97d2ecd7d18188a75da8ea019665818ba067c2ae

**Comment by flyueliu** [comment]
> Thank you very much @flyueliu I've cherry-picked this PR to branch `3_0_X` as [97d2ecd](https://github.com/apache/syncope/commit/97d2ecd7d18188a75da8ea019665818ba067c2ae)

I'm very glad to participate and thank you for your review. I benefited a lot from it. @ilgrosso

**Pull request merged**